src/server/scene/session_container.cpp

```diff
@@ -82,3 +83,27 @@ std::shared_ptr<ms::Session> ms::SessionContainer::successor_of(std::shared_ptr<
 
     BOOST_THROW_EXCEPTION(std::logic_error("Invalid session"));
 }
+
+auto mir::scene::SessionContainer::predecessor_of(std::shared_ptr<Session> const& session) const
+    -> std::shared_ptr<Session>
+{
+    std::shared_ptr<Session> result, first;
+
+    if (!session && apps.size())
+        return apps.front();
+    else if(!session)
+        return std::shared_ptr<Session>();
+
+    for (auto it = apps.rbegin(); it != apps.rend(); it++)
+    {
+        if (*it == session)
+        {
+            auto successor = ++it;
+            if (successor == apps.rend())
+                return *apps.rbegin();
+            else return *successor;
```

Comment:
Like `focus_prev_application`, *maybe* this needs a whole separate implementation from `successor_of`, but I'd like to hear why.

Reply:
There are ways to factor out the difference between `begin()` and `rbegin()` etc. but I don't see that the code would be more reliable or easier to follow.

Reply:
In contrast to `focus_{next,previous}_application()` there's basically no logic here. I don't think there's anything to be gained trying to factor this out.